Ensure final_pipeline is executed once

HiDAl opened the pull request:
In case, a processor changes the index of a request and the new index has a final_pipeline, this pipeline is executed twice. This happens because `IngestService` overwrites the pipeline information, but it never cleans it up later. This change ensures that `IngestService` does not change the pipeline information for this case using methods without side effects


Closes #83653

elasticsearchmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

joegallo reviewed (changes requested):
The change you're making in `executePipelines` is, IMHO, just working around a bug in `resolvePipelines`. That method should handle this, not the caller of that method. Also that'll make it easier to add a unit test, which I would like to see, please (the yaml test is great, though, we should keep that, too).

HiDAl commented:
@joegallo more than a "bug in `resolvePipelines`" I think it's a design problem. I've refactored the code extracting all the pipelines' related logic to its own class and split up the code.

joegallo commented on server/src/main/java/org/elasticsearch/ingest/PipelineUtils.java:
```diff
@@ -0,0 +1,194 @@
+
+public class PipelineUtils {
+
+    public static final String NOOP_PIPELINE_NAME = "_none";
```
I'm not sold on a utils class in general, but regardless of whether we go that way or not, I'm not in favor of it owning this constant -- I think you should put it back on `IngestService`.

HiDAl replied:
I neither, but this `IngestService` class is a bit bloated with a lot of logic, and doing this method split (into resolve + calculate methods) would increase the needed cognitive load. I'll move back to `IngestService` the logic.

As you know, I'm still in the process of learning elastic's mores.

joegallo commented:
> Closes (#83653)

https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword 

Please try to use the github syntax for this (that is, no parens). Also as a nit, try to save references like that for the description of the PR and not the message of a commit -- if you use it in the message of a commit and you rebase your commits, then the linked issue gets cluttered up with lots of references to extraneous commits.

HiDAl commented:
I've moved back the logic to `IngestService` and also figured out a way to unit test the bug :)

Commit:
Ensure final_pipeline is executed once

In case, a processor changes the index of a request and the new index
has a final_pipeline, this pipeline executed twice. This happens because
`IngestService` overwrites the pipeline information, but it never
cleans it up later. This change ensures that `IngestService` does not
change the pipeline information for this case using methods without side
effects

Closes #83653

HiDAl commented:
PR waiting for merge of #92455

elasticsearchmachine commented:
Hi @HiDAl, I've created a changelog YAML for you.

HiDAl committed:
Update docs/changelog/92400.yaml

joegallo commented:
@HiDAl when you come back to this, I think it might be simpler if you tease this apart into two separate PRs. Either just a bugfix and then just a refactoring, or the reverse. My goal there is that I think two laser-focused PRs might be easier to get merged here than one PR that does both.

HiDAl commented:
Closing in favor of #93821 & #93863
